### ISSUE TITLE
Bug in the code generated by the MAlonzo backend

### ISSUE THREAD

**Issue by GoogleCodeExporter** [agda/agda#1624]
I'm using master (idem for the standard library) compiled with
GHC version 7.10.1.20150612 and I have stumbled upon a bug I did
not encounter last time I compiled my project (AIM XXI).

In Test.agda, I define a function pattern-matching on a
list of Chars like so:

``` agda
===================== Test.agda =========================
module Test where

open import Data.Char
open import Data.String
open import Data.List

test : List Char → String
test ('0' ∷ 'x' ∷ xs) = "x"
test ('0' ∷ 'b' ∷ xs) = "b"
test xs               = "f"
=========================================================
```

I then compile it to Haskell using the command line:

```
===================== Compile ===========================
agda -i . -i ~/PATH_TO_STDLIB --compile Test.agda
=========================================================
```

Which generates the following file (no need to wait for the
97 modules to compile btw).

The interesting thing here is the variable `v2`: two _distinct_
case-analysis are performed on it, one to compare it to the
character `'0'` and one to compare it to a list `(:) v4 v5`.
The second case analysis should be on `v3`.

```
================ MAlonzo/Code/Test.hs ==================
{-# LANGUAGE EmptyDataDecls, ExistentialQuantification,
  ScopedTypeVariables, NoMonomorphismRestriction, Rank2Types #-}
module MAlonzo.Code.Test where
import qualified Data.FFI
import qualified MAlonzo.RTE
import qualified MAlonzo.Code.Data.Char
import qualified MAlonzo.Code.Data.List
import qualified MAlonzo.Code.Data.String
name2 = "Test.test"
d2
  = MAlonzo.RTE.mazCoerce
      (\ v0 ->
         let v1 = "f" in
           case MAlonzo.RTE.mazCoerce v0 of
               (:) v2 v3 -> MAlonzo.RTE.mazCoerce
                              (case MAlonzo.RTE.mazCoerce v2 of
                                   '0' -> MAlonzo.RTE.mazCoerce
                                            (case MAlonzo.RTE.mazCoerce v2 of
                                                 (:) v4 v5 -> MAlonzo.RTE.mazCoerce
                                                                (case MAlonzo.RTE.mazCoerce v4 of
                                                                     'b' -> MAlonzo.RTE.mazCoerce
                                                                              "b"
                                                                     'x' -> MAlonzo.RTE.mazCoerce
                                                                              "x"
                                                                     _ -> v1)
                                                 _ -> v1)
                                   _ -> v1)
               _ -> v1)
=========================================================
```

Original issue reported on code.google.com by `g.x.all...@gmail.com` on 7 Aug 2015 at 4:49

**Comment by phile314** [comment]
Definitely going something wrong here. I suspect that this has to do with the recent Compiler / Treeless / CompiledClauses changes. I will try to look at it later this week.

**Comment by phile314** [comment]
I could reproduce it for the MAlonzo and UHC backend in master. For the UHC backend, the regression is introduced by cc15ec788649d59fa42a8720bf1a51f5bf9ae5fc .

**Comment by phile314** [comment]
It seems that the bug is a problem in the CompiledClause (CC) -> Treeless translation.

CC for the test function:

<pre>
case 0 of
  Common.List.List._∷_ ->
    case 0 of
      LitChar '0' ->
        case 0 of
          Common.List.List._∷_ ->
            case 0 of
              LitChar 'b' -> done[[]r("xs")] Shared #301{Lit (LitString "b")}
              LitChar 'x' -> done[[]r("xs")] Shared #318{Lit (LitString "x")}
  _ -> done[[]r("xs")] Shared #300{Lit (LitString "f")}
</pre>

At the moment, the CC->Treeless translation assumes that CC literal cases don't
grow or shrink the CC environment. This seems wrong, and it looks like doing "case x of .."
always removes the case scrutinee x from the CC environment.

**Comment by phile314** [comment]
Fixed by 77749d0 .
